onnx/onnx.in.proto

```diff
@@ -196,6 +204,117 @@ message NodeProto {
   optional string doc_string = 6;
 }
 
+// Training information
+// TrainingInfoProto stores information for training a model.
+// In particular, this defines two functionalities: an initialization-step
+// and a training-algorithm-step. The semantics of the initialization-step is
+// that all initializers (in ModelProto.graph and in all
+// TrainingInfoProto.algorithm) are first initialized as specified by the
+// initializers in the graph, and then by the "initialization_binding" in every
+// instance in ModelProto.training_info. The semantics of the
+// training-algorithm-step is that all the initializers are updated as
+// specified by the "update_binding" in every instance in training_info.
+//
+// The field "algorithm" defines a computation graph similar to the inference
+// one (i.e., ModelProto.graph) but represents a training algorithm's step.
+// Each execution of a TrainingInfoProto.algorithm leads to the update of some
+// initializers in its TrainingInfoProto.algorithm.initializer and
+// ModelProto.graph.initializer.
+//
+// If there are multiple TrainingInfoProto's, sequentially executing all of
+// them once constitutes a full training algorithm. The initializer update
+// happens at the end of each TrainingInfoProto's execution. If the targeted
+// training algorithm contains consecutive update stages (such as block
+// coordinate descent methods), the user needs to create a TrainingInfoProto
+// for each stage.
+//
+// In addition to training (improving the model based on input data),
+// TrainingInfoProto can store an initialization process (resetting the model
+// back to its original state as if no training has been consumed) as a
+// computation graph in the "initialization" field. The execution of
+// "initialization" is similar to that of "algorithm". The user can use
+// "initialization_binding" to assign the outputs of "initialization" to
+// ModelProto.graph.initializer and TrainingInfoProto.algorithm.initializer.
+// 
+// Several inference components are usable in TrainingInfoProto; the list
+// includes
+//   1. ModelProto.graph is a callable inside TrainingInfoProto.algorithm.
+//      Each call is a NodeProto in TrainingInfoProto.algorithm.node with
+//      NodeProto.op_type == "Graph" and
+//      NodeProto.name == ModelProto.graph.name.
```

Comment:
By using the name as the graph identifier, you will only be able to have one instance, since node names should be unique. I think it would be better to use an attribute, e.g. `graph_name`.

Reply:
I will switch to `NodeProto.op_type == ModelProto.graph.name` as suggested by @gramalingam.  This strategy implies callable graphs can't have names conflicting with existing ONNX operators.

Reply:
That conflates names from the `Graph` namespace and the `Operator` namespace. cref https://github.com/onnx/onnx/blob/master/docs/IR.md#names-within-a-graph

I think the `Graph` operator is a much clearer mechanism:
```
{ op_type = "Graph",
  domain = "", # Default ai.onnx domain
  name = "node_name_used_for_diagnostics_only",
  attribute = {
    "graph_name" = <ModelProto.graph.name>,
  },
}
```